mesa/visualization/components/altair_components.py

```diff
@@ -168,23 +180,238 @@ def _draw_grid(space, agent_portrayal):
+        chart_width: width of the agent chart to maintain consistency with the property charts
+        chart_height: height of the agent chart to maintain consistency with the property charts
+    Returns:
+        Altair Chart
+    """
+    try:
+        # old style spaces
+        property_layers = space.properties
+    except AttributeError:
+        # new style spaces
+        property_layers = space._mesa_property_layers
+    base = None
+    for layer_name, portrayal in propertylayer_portrayal.items():
+        layer = property_layers.get(layer_name, None)
+        if not isinstance(
+            layer,
+            PropertyLayer | mesa.discrete_space.property_layer.PropertyLayer,
+        ):
+            continue
+
+        data = layer.data.astype(float) if layer.data.dtype == bool else layer.data
+
+        if (space.width, space.height) != data.shape:
+            warnings.warn(
+                f"Layer {layer_name} dimensions ({data.shape}) do not match space dimensions ({space.width}, {space.height}).",
+                UserWarning,
+                stacklevel=2,
+            )
+        alpha = portrayal.get("alpha", 1)
+        vmin = portrayal.get("vmin", np.min(data))
+        vmax = portrayal.get("vmax", np.max(data))
+        colorbar = portrayal.get("colorbar", True)
+
+        # Prepare data for Altair (convert 2D array to a long-form DataFrame)
+        df = pd.DataFrame(
+            {
+                "x": np.repeat(np.arange(data.shape[0]), data.shape[1]),
+                "y": np.tile(np.arange(data.shape[1]), data.shape[0]),
+                "value": data.flatten(),
+            }
+        )
+
+        if "color" in portrayal:
+            # any value less than vmin will be mapped to the color corresponding to vmin
+            # any value more than vmax will be mapped to the color corresponding to vmax
+            def apply_rgba(val, vmin=vmin, vmax=vmax, alpha=alpha, portrayal=portrayal):
+                a = (val - vmin) / (vmax - vmin)
+                a = max(0, min(a, 1))  # to ensure that a is between 0 and 1
+                a *= alpha  # vmax will have an opacity corresponding to alpha
+                rgb_color = to_rgb(portrayal["color"])
+                r = int(rgb_color[0] * 255)
+                g = int(rgb_color[1] * 255)
+                b = int(rgb_color[2] * 255)
+
+                return f"rgba({r}, {g}, {b}, {a:.2f})"
+
+            df["color"] = df["value"].apply(apply_rgba)
+
+            chart = (
+                alt.Chart(df)
+                .mark_rect()
+                .encode(
+                    x=alt.X("x:O", axis=None),
+                    y=alt.Y("y:O", axis=None),
+                    fill=alt.Fill("color:N", scale=None),
+                )
+                .properties(width=chart_width, height=chart_height, title=layer_name)
+            )
+            base = (base + chart) if base is not None else chart
```

Comment:
You are overcomplicating it a bit, this should work:

```suggestion
import matplotlib.colors as mcolors
if "color" in portrayal:
    color = portrayal["color"]
    
    # Convert the user color + alpha to RGBA
    rgba = mcolors.to_rgba(color, alpha=alpha)
    
    layer_chart = (
        alt.Chart(df)
        .mark_rect(
            color=f"rgba({int(rgba[0] * 255)}, "
                  f"{int(rgba[1] * 255)}, "
                  f"{int(rgba[2] * 255)}, "
                  f"{rgba[3]})"
        )
        .encode(
            x=alt.X("x:O", axis=None),
            y=alt.Y("y:O", axis=None),
            opacity=alt.Opacity(
                "value:Q", 
                scale=alt.Scale(domain=[vmin, vmax], range=[0, 1])
            )
        )
        .properties(width=chart_width, height=chart_height, title=layer_name)
    )
```


Colorbar should be implemented in altair as well, though I am also not very sure how will that work, I have been trying for the past hour to get it right, but either its overcomplicating the code or its not working.

Reply:
I actually tried this out, but I am not able to create a color bar when I am using the code you have suggested as I don't know the mapping of value to color when the Scale function is used and the inbuilt altair legend is discrete and not continous

Reply:
Yeah, but the matplotlib implementation of just the colorbar is also not right.